suuchi-core/src/main/scala/in/ashwanthkumar/suuchi/store/VersionedStore.scala

```diff
@@ -0,0 +1,87 @@
+package in.ashwanthkumar.suuchi.store
+
+import in.ashwanthkumar.suuchi.store.Store
+import in.ashwanthkumar.suuchi.utils.DateUtils
+
+import scala.util.hashing.MurmurHash3
+
+
+object Versions {
+  def fromString(versionString: String) = versionString.split('|').map(_.toLong)
+  def toString(versions: List[Long]) = versions.map(_.toString).mkString("|")
+}
+
+object VersionedStore {
+  val VERSION_PREFIX = "V_"
+  val DATA_PREFIX = "D_"
+  def vkey(key: Array[Byte]) = VERSION_PREFIX.getBytes ++ key
+  def dkey(key: Array[Byte], version: Long) = DATA_PREFIX.getBytes ++ key ++ version.toString.getBytes
+}
+
+class VersionedStore(store: Store, numVersions: Int) extends Store with DeleteStore with DateUtils {
+  import VersionedStore._
+  val SYNC_SLOTS = Array.fill(8192)(new Object)
+
+  override def get(key: Array[Byte]) : Option[Array[Byte]] = {
+    // fetch version record
+    val vRecord = store.get(vkey(key))
+    if(vRecord.isEmpty) None
+    else {
+      val versions = Versions.fromString(new String(vRecord.get))
+      get(key, versions.max)
+    }
+  }
+
+  def get(key: Array[Byte], version: Long) : Option[Array[Byte]] = {
+    store.get(dkey(key, version))
+  }
+
+  override def put(key: Array[Byte], value: Array[Byte]): Boolean = {
+    val nowMs = now
+
+    // atomically update version metadata
+    val versions = atomicUpdate(key, nowMs)
+
+    // remove oldest version, if we've exceeded max # of versions per record
```

Comment:
Might be too much and I know you're going to scold me for this, but still 😛  - do you think it's worthwhile to off-load this task on a thread and forget about it? That way the caller is not blocked for the purge to happen.